provider/pkg/await/internal/awaiter.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO:
+// event aggregator
+// thing to spawn a bunch of observers
+// legacy condition to bridge shit
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/pulumi/pulumi-kubernetes/provider/v4/pkg/await/condition"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// Awaiter orchestrates a condition Satisfier and optional Observers.
+type Awaiter struct {
+	namespace string
+	condition condition.Satisfier
+	observers []condition.Observer
+}
+
+// NewAwaiter creates a new Awaiter with the given options.
+func NewAwaiter(options ...awaiterOption) (*Awaiter, error) {
+	ea := &Awaiter{}
+	for _, opt := range options {
+		opt.apply(ea)
+	}
+	return ea, nil
+}
+
+// Await blocks until the Condition is met or until the context is closed.
+// The operation's timeout should be applied to the provided Context.
+func (ea *Awaiter) Await(ctx context.Context) (err error) {
+	if ea.condition == nil {
+		return fmt.Errorf("missing condition")
+	}
+
+	// check channel is signaled when we should re-evaluate our condition.
+	check := make(chan struct{})
+
+	// We'll spawn a goroutine for our condition and each observer. We wait for
+	// them to tear down before returning because the condition might change
+	// during that time.
+	wg := sync.WaitGroup{}
+	wg.Add(len(ea.observers) + 1)
+	go func() {
+		wg.Wait()
+		close(check)
+	}()
+
+	// Start all of our observers.
+	observers := append([]condition.Observer{ea.condition}, ea.observers...)
+	for _, o := range observers {
+		go func(o condition.Observer) {
+			defer wg.Done()
+			o.Range(func(e watch.Event) bool {
+				_ = o.Observe(e)
+				// Re-evaluate our condition if we see an event for it.
+				if _, ok := o.(condition.Satisfier); ok {
+					check <- struct{}{}
+				}
+				return true
+			})
+		}(o)
+	}
+
+	// Before returning we attempt to re-evaluate the condition a final time,
+	// and we wrap our error with our object's last known state so it can be
+	// checkpointed.
+	defer func() {
+		if err == nil {
+			// Nothing to do.
+			return
+		}
+		// Make sure Observers are all done.
+		wg.Wait()
```

Comment:
In the event of context deadline being exceeded, wouldn't this mean that the awaiter still continues waiting/observing? Shouldn't we just skip waiting for the wait group to be done?

Reply:
The context is plumbed all the way down and respected by the observers, so if it dies all of our informers will shut down and this will resolve. Namely these guys:

https://github.com/pulumi/pulumi-kubernetes/pull/3133/files#diff-c68cee828d9c5172eef833ba32b6185741c858ff507bd2f4d6df8c5a6fb275dbR55-R59

https://github.com/pulumi/pulumi-kubernetes/pull/3133/files#diff-b52d2594c41dd1ff41784e1c6101fcd3c86f51c128ea8b8de43ddc6607977a15R148

Full disclosure I'm very skeptical we need any of this "Hail Mary" logic. There's a comment in the code but I think it largely stems from issues we had handling watch errors. I kept it as-is since we've got tests around it, but I expect we could get rid of it without issue.

Reply:
I've simplified this a bit -- we don't need to worry about context cancellation here since the observer's already shut down when that happens.